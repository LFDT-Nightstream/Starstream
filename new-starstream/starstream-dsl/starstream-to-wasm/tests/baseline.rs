use starstream_to_wasm::compile;
use starstream_types::{BinaryOp, Block, Expr, Identifier, Literal, Program, Spanned, Statement};

macro_rules! assert_wat_snapshot {
    ($program:expr) => {{
        let program = $program;
        let wasm = compile(&program);
        let mut wat = String::new();
        wasmprinter::Config::new().fold_instructions(true).print(
            &wasm,
            &mut wasmprinter::PrintFmtWrite(&mut wat),
        ).unwrap();
        insta::with_settings!({
            description => format!("{:?}", program),
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(wat);
        });
    }}
}

#[test]
fn empty() {
    assert_wat_snapshot!(&Program { statements: vec![] });
}

#[test]
fn simple_while_loop() {
    assert_wat_snapshot!(&Program {
        statements: vec![
            Statement::VariableDeclaration {
                name: Identifier::new("foo", None),
                value: Spanned::none(Expr::Literal(Literal::Integer(0))),
            },
            Statement::While {
                condition: Spanned::none(Expr::Binary {
                    op: BinaryOp::Less,
                    left: Box::new(Spanned::none(Expr::Identifier(Identifier::new(
                        "foo", None
                    )))),
                    right: Box::new(Spanned::none(Expr::Literal(Literal::Integer(10)))),
                }),
                body: Block {
                    statements: vec![Statement::Assignment {
                        target: Identifier::new("foo", None),
                        value: Spanned::none(Expr::Binary {
                            op: BinaryOp::Add,
                            left: Box::new(Spanned::none(Expr::Identifier(Identifier::new(
                                "foo", None
                            )))),
                            right: Box::new(Spanned::none(Expr::Literal(Literal::Integer(1)))),
                        }),
                    }],
                },
            },
        ],
    });
}
