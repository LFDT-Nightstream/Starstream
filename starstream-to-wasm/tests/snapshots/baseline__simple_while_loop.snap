---
source: starstream-to-wasm/tests/baseline.rs
description: "Program { statements: [VariableDeclaration { name: Identifier { name: \"foo\", span: None }, value: Spanned { node: Literal(Integer(0)), span: 0..0 } }, While { condition: Spanned { node: Binary { op: Less, left: Spanned { node: Identifier(Identifier { name: \"foo\", span: None }), span: 0..0 }, right: Spanned { node: Literal(Integer(10)), span: 0..0 } }, span: 0..0 }, body: Block { statements: [Assignment { target: Identifier { name: \"foo\", span: None }, value: Spanned { node: Binary { op: Add, left: Spanned { node: Identifier(Identifier { name: \"foo\", span: None }), span: 0..0 }, right: Spanned { node: Literal(Integer(1)), span: 0..0 } }, span: 0..0 } }] } }] }"
---
(module
  (type (;0;) (func))
  (export "main" (func 0))
  (func (;0;) (type 0)
    (local i64)
    (local.set 0
      (i64.const 0))
    (block ;; label = @1
      (loop ;; label = @2
        (br_if 1 (;@1;)
          (i32.eqz
            (i64.lt_s
              (local.get 0)
              (i64.const 10))))
        (local.set 0
          (i64.add
            (local.get 0)
            (i64.const 1)))
        (br 0 (;@2;))))
  )
)
